### PR TITLE
Add sample dataset option in `v1/nsql`.

### PR DESCRIPTION
**Pull request by Jeadie**
## 🗣 Description
 - Allow users to sample data when using `v1/nsql`.

### Example
```shell
curlie -XPOST -H "Content-Type: application/json" "http://localhost:8090/v1/nsql" -d '{
  "query": "What is the total sales for the month of January 2020?",
  "datasets": ["sales"], // ommitting includes all datasets
  "sample_data_enabled": true | false, // default true.
}'
```

Chat Completion message
```json
{
  "messages":   "messages": [
    {
      "role": "assistant",
      "tool_calls": [
        {
          "id": "initial_list_datasets",
          "type": "function",
          "function": {
            "name": "list_datasets",
            "arguments": ""
          }
        }
      ]
    },
    {
      "role": "tool",
      "content": "[{\"can_search_documents\":false,\"description\":null,\"metadata\":{},\"table\":\"spice.public.sales\"}]",
      "tool_call_id": "initial_list_datasets"
    },
    {
      "role": "system",
      "content": "```SQL\nCREATE TABLE IF NOT EXISTS \"sales\" ( \"order_number\" bigint, \"quantity_ordered\" bigint, \"price_each\" double precision, \"order_line_number\" bigint, \"sales\" double precision, \"order_date\" timestamp, \"status\" text, \"quarter\" bigint, \"month\" bigint, \"year\" bigint, \"product_line\" text, \"msrp\" bigint, \"product_code\" text, \"customer_name\" text, \"phone\" text, \"address_line1\" text, \"address_line2\" text, \"city\" text, \"state\" text, \"postal_code\" text, \"country\" text, \"territory\" text, \"contact_last_name\" text, \"contact_first_name\" text, \"deal_size\" text )```\nTask: Write a postgres SQL query to answer this question: _\"What is the total sales for the month of January 2020?\"_. Instruction: Return only valid SQL code, nothing additional."
    },
    {
      "role": "assistant",
      "tool_calls": [
        {
          "id": "distinct-sales-nsql",
          "type": "function",
          "function": {
            "name": "sample_data",
            "arguments": "{\"dataset\":\"sales\",\"limit\":3,\"cols\":null}"
          }
        }
      ]
    },
    {
      "role": "tool",
      "content": "\"+--------------+------------------+------------+-------------------+--------+---------------------+------------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-----------------------+---------------+-----------+-------+-------------+-----------+-----------+-------------------+--------------------+-----------+\\n| order_number | quantity_ordered | price_each | order_line_number | sales  | order_date          | status     | quarter | month | year | product_line | msrp | product_code | customer_name      | phone            | address_line1         | address_line2 | city      | state | postal_code | country   | territory | contact_last_name | contact_first_name | deal_size |\\n+--------------+------------------+------------+-------------------+--------+---------------------+------------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-----------------------+---------------+-----------+-------+-------------+-----------+-----------+-------------------+--------------------+-----------+\\n| 10100        | 6                | 26.88      | 1                 | 482.13 | 2003-01-06T00:00:00 | Cancelled  | 1       | 1     | 2003 | Classic Cars | 33   | S10_1678     | AV Stores, Co.     | (02) 5554 67     | 1 rue Alsace-Lorraine | 2nd Floor     | Aaarhus   | BC    | 10022       | Australia | APAC      | Accorti           | Adrian             | Large     |\\n| 10101        | 10               | 27.22      | 2                 | 541.14 | 2003-01-09T00:00:00 | Disputed   | 2       | 2     | 2004 | Motorcycles  | 35   | S10_1949     | Alpha Cognac       | (071) 23 67 2555 | 11328 Douglas Av.     | Floor No. 4   | Allentown | CA    | 10100       | Austria   | EMEA      | Ashworth          | Akiko              | Medium    |\\n| 10102        | 11               | 28.29      | 3                 | 553.95 | 2003-01-10T00:00:00 | In Process | 3       | 3     | 2005 | Planes       | 37   | S10_2016     | Amica Models & Co. | (1) 42.34.2555   | 12, rue des Bouchers  | Level 15      | Barcelona | CT    | 106-0032    | Belgium   | Japan     | Barajas           | Allen              | Small     |\\n+--------------+------------------+------------+-------------------+--------+---------------------+------------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-----------------------+---------------+-----------+-------+-------------+-----------+-----------+-------------------+--------------------+-----------+\"",
      "tool_call_id": "distinct-sales-nsql"
    },
    {
      "role": "assistant",
      "tool_calls": [
        {
          "id": "distinct-sales-nsql",
          "type": "function",
          "function": {
            "name": "sample_data",
            "arguments": "{\"dataset\":\"sales\",\"limit\":3}"
          }
        }
      ]
    },
    {
      "role": "tool",
      "content": "\"+--------------+------------------+------------+-------------------+---------+---------------------+---------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-------------------------------+---------------+-------+-------+-------------+---------+-----------+-------------------+--------------------+-----------+\\n| order_number | quantity_ordered | price_each | order_line_number | sales   | order_date          | status  | quarter | month | year | product_line | msrp | product_code | customer_name      | phone            | address_line1                 | address_line2 | city  | state | postal_code | country | territory | contact_last_name | contact_first_name | deal_size |\\n+--------------+------------------+------------+-------------------+---------+---------------------+---------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-------------------------------+---------------+-------+-------+-------------+---------+-----------+-------------------+--------------------+-----------+\\n| 10107        | 30               | 95.7       | 2                 | 2871.0  | 2003-02-24T00:00:00 | Shipped | 1       | 2     | 2003 | Motorcycles  | 95   | S10_1678     | Land of Toys Inc.  | 2125557818       | 897 Long Airport Avenue       |               | NYC   | NY    | 10022       | USA     |           | Yu                | Kwai               | Small     |\\n| 10121        | 34               | 81.35      | 5                 | 2765.9  | 2003-05-07T00:00:00 | Shipped | 2       | 5     | 2003 | Motorcycles  | 95   | S10_1678     | Reims Collectables | 26.47.1555       | 59 rue de l'Abbaye            |               | Reims |       | 51100       | France  | EMEA      | Henriot           | Paul               | Small     |\\n| 10134        | 41               | 94.74      | 2                 | 3884.34 | 2003-07-01T00:00:00 | Shipped | 3       | 7     | 2003 | Motorcycles  | 95   | S10_1678     | Lyon Souveniers    | +33 1 46 62 7555 | 27 rue du Colonel Pierre Avia |               | Paris |       | 75508       | France  | EMEA      | Da Cunha          | Daniel             | Medium    |\\n+--------------+------------------+------------+-------------------+---------+---------------------+---------+---------+-------+------+--------------+------+--------------+--------------------+------------------+-------------------------------+---------------+-------+-------+-------------+---------+-----------+-------------------+--------------------+-----------+\"",
      "tool_call_id": "distinct-sales-nsql"
    }
  ],
  "model": "nql",
  "response_format": {
    "type": "json_schema",
    "json_schema": {
      "name": "sql_mode",
      "schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": false,
        "properties": {
          "sql": {
            "type": "string"
          }
        },
        "required": [
          "sql"
        ],
        "title": "StructuredNsqlOutput",
        "type": "object"
      },
      "strict": true
    }
  }
}
```

### CODE REVIEW

**Comment by Jeadie** [comment]
Will wait for #3108  to rejig this.

File: crates/runtime/src/tools/builtin/sample/mod.rs [hunk 1]
```diff
@@ -84,6 +87,36 @@ pub enum SampleTableParams {
...
+impl From<&SampleTableParams> for SampleTableMethod {
+    fn from(params: &SampleTableParams) -> Self {
+        match params {
+            SampleTableParams::DistinctColumns(_) => SampleTableMethod::DistinctColumns,
+            SampleTableParams::RandomSample(_) => SampleTableMethod::RandomSample,
+            SampleTableParams::TopNSample(_) => SampleTableMethod::TopNSample,
+        }
+    }
+}
+impl SampleTableParams {
```
Comment (phillipleblanc) [inline crates/runtime/src/tools/builtin/sample/mod.rs]:
```suggestion
}

impl SampleTableParams {
```

File: crates/runtime/src/http/v1/nsql.rs [hunk 2]
```diff
@@ -42,13 +61,96 @@ fn clean_model_based_sql(input: &str) -> String {
...
     pub query: String,
 
     #[serde(default = "default_model")]
     pub model: String,
+
+    #[serde(default = "default_sample_data_enabled")]
+    pub sample_data_enabled: bool,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub datasets: Option<Vec<String>>,
```
Comment (peasee) [inline crates/runtime/src/http/v1/nsql.rs:149]: Do we expect users to be able to exclude this property entirely in their requests? If so, this needs a `serde(default, ...)`

Comment (Jeadie) [reply crates/runtime/src/http/v1/nsql.rs:149]: I don't think so?
```shell
curlie -XPOST -H "Content-Type: application/json" "http://localhost:8090/v1/nsql" -d '{
  "query": "What is the total sales?"
}'
```

**Review by lukekim — APPROVED** [review]

**Pull request merged**